File: blimp/client/core/contents/android/java/src/org/chromium/blimp/core/contents/BlimpNavigationControllerImpl.java
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.blimp.core.contents;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.blimp_public.contents.BlimpNavigationController;

/**
 * BlimpNavigationControllerImpl is a Java wrapper to allow communicating with the native
 * BlimpNavigationControllerImpl object.
 */
@JNINamespace("blimp::client")
public class BlimpNavigationControllerImpl implements BlimpNavigationController {
    @CalledByNative
    private static BlimpNavigationControllerImpl create(
            long nativeBlimpNavigationControllerImplAndroid) {
        return new BlimpNavigationControllerImpl(nativeBlimpNavigationControllerImplAndroid);
    }

    private long mNativeBlimpNavigationControllerImplAndroid;

    private BlimpNavigationControllerImpl(long blimpNavigationControllerImplAndroid) {
        mNativeBlimpNavigationControllerImplAndroid = blimpNavigationControllerImplAndroid;
    }

    @CalledByNative
    private void clearNativePtr() {
        mNativeBlimpNavigationControllerImplAndroid = 0;
    }

    @CalledByNative
    private long getNativePtr() {
        assert mNativeBlimpNavigationControllerImplAndroid != 0;
        return mNativeBlimpNavigationControllerImplAndroid;
    }

    @Override
    public void loadUrl(String url) {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return;
        nativeLoadURL(mNativeBlimpNavigationControllerImplAndroid, url);
    }

    @Override
    public String getUrl() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return "";
        return nativeGetURL(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public String getTitle() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return "";
        return nativeGetTitle(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public boolean canGoBack() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return false;
        return nativeCanGoBack(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public boolean canGoForward() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return false;
        return nativeCanGoForward(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public void goBack() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return;
        nativeGoBack(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public void goForward() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return;
        nativeGoForward(mNativeBlimpNavigationControllerImplAndroid);
    }

    @Override
    public void reload() {
        if (mNativeBlimpNavigationControllerImplAndroid == 0) return;
        nativeReload(mNativeBlimpNavigationControllerImplAndroid);
    }

    private native void nativeGoBack(long nativeBlimpNavigationControllerImplAndroid);
    private native void nativeGoForward(long nativeBlimpNavigationControllerImplAndroid);
    private native void nativeReload(long nativeBlimpNavigationControllerImplAndroid);
    private native boolean nativeCanGoBack(long nativeBlimpNavigationControllerImplAndroid);
    private native boolean nativeCanGoForward(long nativeBlimpNavigationControllerImplAndroid);
    private native void nativeLoadURL(long nativeBlimpNavigationControllerImplAndroid, String url);
    private native String nativeGetURL(long nativeBlimpNavigationControllerImplAndroid);
    private native String nativeGetTitle(long nativeBlimpNavigationControllerImplAndroid);
}

File: blimp/client/core/contents/blimp_navigation_controller_impl.cc
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "blimp/client/core/contents/blimp_navigation_controller_impl.h"

#include "blimp/client/core/contents/blimp_navigation_controller_delegate.h"

namespace {
// TODO(shaktisahu): NavigationFeature currently needs a tab_id. Remove this
// later when it is fully integrated with BlimpClientContext.
const int kDummyTabId = 0;
}  // namespace

namespace blimp {
namespace client {

BlimpNavigationControllerImpl::BlimpNavigationControllerImpl(
    BlimpNavigationControllerDelegate* delegate,
    NavigationFeature* feature)
    : navigation_feature_(feature), delegate_(delegate) {
  if (navigation_feature_)
    navigation_feature_->SetDelegate(kDummyTabId, this);
}

BlimpNavigationControllerImpl::~BlimpNavigationControllerImpl() = default;

void BlimpNavigationControllerImpl::SetNavigationFeatureForTesting(
    NavigationFeature* feature) {
  navigation_feature_ = feature;
  navigation_feature_->SetDelegate(kDummyTabId, this);
}

void BlimpNavigationControllerImpl::LoadURL(const GURL& url) {
  current_url_ = url;
  navigation_feature_->NavigateToUrlText(kDummyTabId, current_url_.spec());
}

void BlimpNavigationControllerImpl::Reload() {
  navigation_feature_->Reload(kDummyTabId);
}

bool BlimpNavigationControllerImpl::CanGoBack() const {
  NOTIMPLEMENTED();
  return true;
}

bool BlimpNavigationControllerImpl::CanGoForward() const {
  NOTIMPLEMENTED();
  return true;
}

void BlimpNavigationControllerImpl::GoBack() {
  navigation_feature_->GoBack(kDummyTabId);
}

void BlimpNavigationControllerImpl::GoForward() {
  navigation_feature_->GoForward(kDummyTabId);
}

const GURL& BlimpNavigationControllerImpl::GetURL() {
  return current_url_;
}

const std::string& BlimpNavigationControllerImpl::GetTitle() {
  return current_title_;
}

void BlimpNavigationControllerImpl::OnUrlChanged(int tab_id, const GURL& url) {
  current_url_ = url;
  delegate_->OnNavigationStateChanged();
}

void BlimpNavigationControllerImpl::OnFaviconChanged(int tab_id,
                                                     const SkBitmap& favicon) {
  delegate_->OnNavigationStateChanged();
}

void BlimpNavigationControllerImpl::OnTitleChanged(int tab_id,
                                                   const std::string& title) {
  current_title_ = title;
  delegate_->OnNavigationStateChanged();
}

void BlimpNavigationControllerImpl::OnLoadingChanged(int tab_id, bool loading) {
  delegate_->OnNavigationStateChanged();
}

void BlimpNavigationControllerImpl::OnPageLoadStatusUpdate(int tab_id,
                                                           bool completed) {
  delegate_->OnNavigationStateChanged();
}

}  // namespace client
}  // namespace blimp
